File: 202.happy-number.java
/*
 * @lc app=leetcode id=202 lang=java
 *
 * [202] Happy Number
 */

// @lc code=start
class Solution {
    public boolean isHappy(int n) {

        int fast = n;
        int slow =n;
        do{
            fast = squaredDigitSum(squaredDigitSum(fast));
            slow = squaredDigitSum(slow);
            if(fast==1 ){
                return true;
            }
        }while(slow!=fast);

        return false;
        
    }
    public int squaredDigitSum(int n){
        int sum=0;
        int d =n;
        while(n!=0){
            sum += (n%10)*(n%10);
            n=n/10;

        }
        // System.out.println(d+" : "+sum);
        return sum;

    }
}
// @lc code=end

File: 202.happy-number.py
#
# @lc app=leetcode id=202 lang=python3
#
# [202] Happy Number
#

# @lc code=start
from typing import Dict


class Solution:
    def isHappy(self, n: int) -> bool:
        fast = n
        slow= n
        f = 0
        while(f==0 or slow!=fast):
            f=1
            fast = self.sumOfSquare(self.sumOfSquare(fast))
            if fast==1:
                return True
            slow = self.sumOfSquare(slow)

        return False
    
    def sumOfSquare(self, n:int):
        return sum(int(i)**2 for i in str(n))
        
        
# @lc code=end

